core/utils_test.go

```diff
@@ -0,0 +1,133 @@
+package core_test
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/hyperledger-labs/yui-relayer/core"
+)
+
+func TestRunUntilComplete(t *testing.T) {
+	runtimeError := errors.New("runtime error")
+
+	tests := []struct {
+		name    string
+		fn      func() (bool, error)
+		attempt int
+		cancel  bool
+		err     error
+	}{
+		{
+			name: "Complete immediately",
+			fn: func() (bool, error) {
+				return true, nil
+			},
+			attempt: 1,
+			cancel:  false,
+			err:     nil,
+		},
+		{
+			name: "Complete on the second try",
+			fn: func() func() (bool, error) {
+				attempt := 0
```

Comment:
Why nested lambda functions here?
Since the outermost function is immediately called here, this statement is simply equal to `fn: <innermost_func>`, right?

Reply:
This is a closure and I use it to keep the call count (`attempt`). I'm afraid that we cannot do this without a closure.

Reply:
It may be better to receive `attempt` as an argument. I will try it.

Reply:
Oh my sorry, the original code also made sense.
The refactored version is also good, thank you.